scipy/stats/_stats_mstats_common.py

```diff
@@ -426,35 +426,4 @@ def siegelslopes(y, x=None, method="hierarchical"):
     >>> plt.show()
 
     """
-    if method not in ['hierarchical', 'separate']:
-        raise ValueError("method can only be 'hierarchical' or 'separate'")
-    y = np.asarray(y).ravel()
-    if x is None:
-        x = np.arange(len(y), dtype=float)
-    else:
-        x = np.asarray(x, dtype=float).ravel()
-        if len(x) != len(y):
-            raise ValueError("Incompatible lengths ! (%s<>%s)" %
-                             (len(y), len(x)))
```

Comment:
The input validation up to here is best left in Python. That way the signatures and generated C++ code are significantly simpler, and this part of the function is very fast.

After this point you also know that `x` is a float array, so the annotation can simply be `float[:]` instead of `int[:] or float[:] or None`.